servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentUtils.java

```diff
@@ -31,11 +31,46 @@
 
     public static final int CONCURRENT_IDLE = 0;
     public static final int CONCURRENT_EMITTING = 1;
+    private static final int CONCURRENT_PENDING = 2;
 
     private ConcurrentUtils() {
         // No instances.
     }
 
+    /**
+     * Acquire a lock that is exclusively held with no re-entry, but attempts to acquire the lock while it is
+     * held can be detected by {@link #releasePendingLock(AtomicIntegerFieldUpdater, Object)}.
+     * @param lockUpdater The {@link AtomicIntegerFieldUpdater} used to control the lock state.
+     * @param owner The owner of the lock object.
+     * @param <T> The type of object that owns the lock.
+     * @return {@code true} if the lock was acquired, {@code false} otherwise.
+     */
+    public static <T> boolean acquirePendingLock(AtomicIntegerFieldUpdater<T> lockUpdater, T owner) {
```

Comment:
@NiteshKant - note these are utilities that are also shared with https://github.com/apple/servicetalk/pull/1011 and https://github.com/apple/servicetalk/pull/1014 ... which ever PR lands first will bring them in and I will rebase to avoid duplication.